### PR TITLE
Update finetuning parameters and improve data splitting logic

### PR DESCRIPTION
**Pull request by anuragg1209**
## Motivation and Context

- Retuned `examples/finetune_classifier.py` (Higgs) and `examples/finetune_regressor.py` (California housing): LR 1e-5 with `lr_warmup_only=True`, `NUM_EPOCHS=100` with `EARLY_STOPPING_PATIENCE=15`, ctx+query 50k (classifier) / 5k (regressor), and inference subsampling set to the full training set (the previous 50k default is a v2.5 legacy that silently caps context on large datasets).
- Introduced `_ratio_ctx_query_split` so the context/query split is sized per chunk instead of using a fixed absolute query size, which crashed on legitimate remainder chunks smaller than the configured query size.
- `validation_split_ratio=None/0` now cleanly disables validation: training uses all data, per-epoch evaluation is skipped, and early stopping is downgraded with a log message.
- The "best model" concept (in-memory restore and `checkpoint_best.pth`) now exists only when early stopping is active; without it, the last epoch is the result and only interval checkpoints are saved, recording train loss instead of placeholder validation metrics.

## Performance of the example scripts

`examples/finetune_classifier.py` (Higgs):

|  | Test ROC AUC ↑ | Test Log Loss ↓ |
|---|---|---|
| Default TabPFN | 0.8247 | 0.5110 |
| **Finetuned TabPFN** | **0.8322** | **0.5003** |

`examples/finetune_regressor.py` (California housing):

|  | Test MSE ↓ | Test R² ↑ |
|---|---|---|
| Default TabPFN | 0.1350 | 0.9018 |
| **Finetuned TabPFN** | **0.1328** | **0.9034** |

## Public API Changes

`validation_split_ratio` now accepts `None`/`0` to disable validation; `early_stopping` is documented to also control best-checkpoint saving and best-weights restoration.

## How Has This Been Tested?

Example scripts run locally on a 4-GPU DDP setup (results above). New/updated unit tests cover the no-validation fit path (evaluation skipped, early stopping downgraded, checkpoint contents), best-checkpoint gating under `early_stopping=False`, and checkpoint resumption.

### CODE REVIEW

File: src/tabpfn/finetuning/finetuned_base.py [hunk 1]
```diff
@@ -193,6 +193,37 @@ def _snapshot_model_state(model: torch.nn.Module) -> dict[str, torch.Tensor]:
...
+    try:
+        return train_test_split(
+            X, y, test_size=test_size, random_state=random_state, stratify=stratify
+        )
+    except ValueError:
+        if stratify is None:
+            raise
+        return train_test_split(
+            X, y, test_size=test_size, random_state=random_state, stratify=None
+        )
```
[the remark this answers is not in the document]
Comment (anuragg1209) [reply src/tabpfn/finetuning/finetuned_base.py:224]: A human would never comment raise this issue. If this issue ever occurs, it's better to raise the error than to avoid it by setting ctx rows to 1 during fine-tuning.

File: src/tabpfn/finetuning/finetuned_base.py [hunk 2]
```diff
@@ -874,15 +906,15 @@ def _ddp_broadcast_primary_metric(metric: float) -> float:
...
             if self.use_lr_scheduler and scheduler is None:
-                steps_per_epoch = len(finetuning_dataloader)
-                if steps_per_epoch == 0:
-                    logger.warning(
-                        "No training batches available; ending training early.",
-                    )
-                    break
-
                 total_steps = steps_per_epoch * self.epochs
                 warmup_steps = int(total_steps * 0.1)
```
[the remark this answers is not in the document]
Comment (anuragg1209) [reply src/tabpfn/finetuning/finetuned_base.py:942]: Fixed this issue.

File: src/tabpfn/finetuning/finetuned_base.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (anuragg1209) [reply src/tabpfn/finetuning/finetuned_base.py:224]: This case won't happen, but if it does happen in some scenario, then it's better to fail and throw the error.

File: src/tabpfn/finetuning/finetuned_base.py [hunk 3]
```diff
@@ -1035,6 +1098,15 @@ def _ddp_broadcast_primary_metric(metric: float) -> float:
             else:
                 primary_metric = self._get_initial_best_metric()
                 eval_result = EvalResult(primary=primary_metric)
```
[the remark this answers is not in the document]
Comment (anuragg1209) [reply src/tabpfn/finetuning/finetuned_base.py:1104]: last-weights-for-inference was already the behavior (no restore happens without validation); the new commit makes the no-best rule explicit and replaces the placeholder metrics in interval checkpoints with the train loss.

File: src/tabpfn/finetuning/finetuned_base.py (hunk 1 above)
Comment (bejaeger) [inline src/tabpfn/finetuning/finetuned_base.py:224]: I'm wondering, could this lead to train elements with very small context? should we rather skip this element entirely in the dataloader?

Comment (anuragg1209) [reply src/tabpfn/finetuning/finetuned_base.py:224]: Small chunks are already skipped upstream: `shuffle_and_chunk_data` drops remainder chunks under `min_chunk_size=2000`, so this splitter only ever sees full-size chunks or remainders > 2000. In worst case, context is 1600 at the default ratio. The only way to get a genuinely small context is when the entire training set (or the configured ctx+query budget) is tiny, and in that case, skipping would discard all the data rather than a degenerate element.

**Review by bejaeger — APPROVED** [review]
LGTM! <br>Just one thing, can you please document in the PR description what the new performance gain is for the examples?

**Comment by anuragg1209** [comment]
> LGTM! Just one thing, can you please document in the PR description what the new performance gain is for the examples?

I've updated the PR description to include the performance gains for the example scripts.

File: src/tabpfn/finetuning/finetuned_base.py (hunk 1 above)
Comment (anuragg1209) [reply src/tabpfn/finetuning/finetuned_base.py:224]: Marking this as resolved. Please feel free to reopen if anything needs to be addressed.
